src/libs/record/record.c

```diff
@@ -0,0 +1,780 @@
+
+	ksDel (addedKeys);
+	ksDel (modifiedOldKeys);
+	ksDel (modifiedNewKeys);
+	ksDel (removedKeys);
+}
+
+/**
+ * Diff and record changes to the KDB instance in @p handle.
+ * This function is mainly intended for use in the recorder plugin.
+ *
+ * @param handle the KDB instance that changes occured on
+ * @param sessionStorageHandle the KDB instances that shall be used to persist the session diff.
+ *                             You can use the same as for @p handle.
+ * @param newKeys the keyset with changed keys
+ * @param parentKey only changes same or below this key will be determined
+ * @param errorKey used for reporting errors and warnings.
+ *                 As usual, they will be found as meta keys attached to this key.
+ * @retval true - changes were recorded successfully
+ * @retfal false - there was an error during recording - see @p errorKey for further details
+ */
+bool elektraRecordRecord (KDB * handle, KDB * sessionStorageHandle, KeySet * newKeys, Key * parentKey, Key * errorKey)
+{
+	if (handle == NULL)
+	{
+		ELEKTRA_SET_INTERFACE_ERROR (errorKey, "NULL pointer passed for KDB handle");
+		return false;
+	}
+
+	if (sessionStorageHandle == NULL)
+	{
+		ELEKTRA_SET_INTERFACE_ERROR (errorKey, "NULL pointer passed for KDB session storage handle");
+		return false;
+	}
+
+	if (newKeys == NULL)
+	{
+		ELEKTRA_SET_INTERFACE_ERROR (errorKey, "NULL pointer passed for new keys");
+		return false;
+	}
+
+	if (parentKey == NULL)
+	{
+		ELEKTRA_SET_INTERFACE_ERROR (errorKey, "NULL pointer passed for parent key");
+		return false;
+	}
+
+	const Key * activeKey = ksLookupByName (handle->global, ELEKTRA_RECORD_CONFIG_ACTIVE_KEY, 0);
+	if (activeKey == NULL)
+	{
+		ELEKTRA_ADD_RESOURCE_WARNINGF (errorKey, "%s called but recording is not enabled.", __func__);
+		// recording is not activated --> do nothing, but still successful
+		return true;
+	}
+
+	Key * recordConfigurationKey = keyNew (ELEKTRA_RECORD_CONFIG_KEY, KEY_END);
+	Key * sessionRecordingKey = keyNew (ELEKTRA_RECORD_SESSION_KEY, KEY_END);
+
+	if (keyIsBelowOrSame (sessionRecordingKey, parentKey) || keyIsBelowOrSame (recordConfigurationKey, parentKey))
+	{
+		// the parent key is either below our session storage or below our configuration storage
+		// we do not want to record changes to the recording tool itself
+		// do nothing, but still successful
+		keyDel (sessionRecordingKey);
+		keyDel (recordConfigurationKey);
+		return true;
+	}
+
+	KeySet * blacklistedKeys = ksNew (4, recordConfigurationKey, keyDup (sessionRecordingKey, KEY_CP_ALL),
+					  keyNew ("system:/elektra/modules", KEY_END), keyNew ("system:/elektra/version", KEY_END), KS_END);
+
+	// Remove all keys that belong to the recording tool or are blacklisted
+	KeySet * toRecord = ksDup (newKeys);
+	for (elektraCursor it = 0; it < ksGetSize (blacklistedKeys); it++)
+	{
+		ksDel (ksCut (toRecord, ksAtCursor (blacklistedKeys, it)));
+	}
+
+	if (ksGetSize (toRecord) == 0 && ksGetSize (newKeys) != 0)
```

Comment:
Is the 2nd condition (`ksGetSize (newKeys) != 0`) really necessary here?
Maybe you could also add the check for an empty KeySet `newKeys` at the beginning of this function (after the NULL-pointer checks).